American option baron adesi whaley

esztermarton opened the pull request:
Implementing Baron-Adesi Whaley approximation of black scholes formula

esztermarton committed:
Implementing Baron-Adesi Whaley approximation of black scholes formula

gavincyi committed:
Add day count convention 30/360

cyrilchim commented on tf_quant_finance/black_scholes/approximations/american_option.py:
```diff
@@ -0,0 +1,377 @@
+  :param s:
+  :param call:
+  :param dtype:
+  :return:
```
Please remove

cyrilchim replied:
please remove

cyrilchim commented on tf_quant_finance/black_scholes/approximations/american_option.py:
```diff
@@ -0,0 +1,377 @@
+  return tf.cond(
+    tf.reduce_all(calculate_eu),
+    lambda: eu_prices,
+    lambda: tf.where(calculate_eu, eu_prices, am_prices))
```
Here both `tensor`s eu_prices and tf.where(calculate_eu, eu_prices, am_prices) are computed (see documentation [here](https://www.tensorflow.org/api_docs/python/tf/cond))

cyrilchim replied:
If this is an intension, you can remove `tf.cond` (which I would recommend doing anyway)

esztermarton replied:
The intention was that am_prices, while the graph can be constructed (which I thought is how cond works), wouldn't be calculated unless there are some that require it, and if all are eu, then no actual computation would happen there, only graph construction.

cyrilchim commented on tf_quant_finance/black_scholes/approximations/american_option_test.py:
```diff
@@ -0,0 +1,166 @@
+from tensorflow.python.framework import test_util  # pylint: disable=g-direct-tensorflow-import
+
+
+print(dir(tff))
```
please remove

cyrilchim commented on tf_quant_finance/black_scholes/approximations/american_option.py:
```diff
@@ -0,0 +1,377 @@
+  is_call_options = None if call else tf.constant(False)
+  sign = 1 if call else -1
+
+  M = 2 * r / sigma ** 2
```
all names should be decapitalized: 'm' insead of `M`. I assume, you are using notations from the paper. Please add a comment that this is indeed the case

cyrilchim commented on tf_quant_finance/black_scholes/approximations/american_option.py:
```diff
@@ -0,0 +1,377 @@
+
+
+def _option_price_put_or_call(sigma, x, t, r, b, call, dtype):
+  """ Calculates q, a for the american option price formula.
```
Not sure what q and a means. Please add nore informative description

cyrilchim commented on tf_quant_finance/black_scholes/approximations/american_option.py:
```diff
@@ -0,0 +1,377 @@
+    s_seed = s_inf + (x - s_inf) * tf.exp(h)
+
+  s_crit, _, _ = implied_vol_newton.newton_root_finder(
+    value_and_gradient, s_seed, dtype=dtype)
```
You need to add `tolerance` and `max_iterations` parameters. Also, the output of american options pricer should include  convergence results. See the [doctring of implied_vol](https://github.com/google/tf-quant-finance/blob/514236ce735ee2d6bdd9fb9981ce79da4dcaa1cd/tf_quant_finance/black_scholes/implied_vol_newton_root.py#L23)

cyrilchim replied:
Also, please pass the arguments by name

cyrilchim commented on tf_quant_finance/black_scholes/approximations/american_option.py:
```diff
@@ -0,0 +1,377 @@
+  s_inf = x / (1 - 1 / q_inf)
+  h = -(sign * b * t + 2 * sigma * tf.sqrt(t)) * sign * (x / (s_inf - x))
+  if call:
+    s_seed = x + (s_inf - x) * (1 - tf.exp(h))
```
You don't need tf.gather_nd above. use tf.where here to build a starting point

cyrilchim commented on tf_quant_finance/black_scholes/approximations/american_option.py:
```diff
@@ -0,0 +1,377 @@
+  K = 1 - tf.exp(- r * t)
+  q_ = _calc_q(N, M, K, call)
+
+  value = lambda s_crit: vanilla_prices.option_price(
+    volatilities=sigma, strikes=x, expiries=t, spots=s_crit, discount_rates=r,
+    cost_of_carries=b, is_call_options=is_call_options, dtype=dtype) + sign * (
+      1 - tf.exp((b - r) * t) * _ncdf(sign * _calc_d1(
+    s_crit, x, sigma, b, t))) * s_crit / q_ - sign * (s_crit - x)
```
You may build the target function with `tf.where` to eliminate tf.gather_nd above. Does it make sense?

cyrilchim commented on tf_quant_finance/black_scholes/approximations/american_option.py:
```diff
@@ -0,0 +1,377 @@
+      return american_call_prices
+
+    # Put option
+    x_put = tf.gather_nd(strikes, put_indices)
```
Let's try to remove gather_nd.

cyrilchim commented on tf_quant_finance/black_scholes/approximations/american_option.py:
```diff
@@ -0,0 +1,377 @@
+    b_put = tf.gather_nd(cost_of_carries, put_indices)
+
+    q1, a1, s_crit_put = _option_price_put_or_call(
+      sigma_put, x_put, t_put, r_put, b_put, False, dtype)
```
I think this function can be rewritten with tf.where without any need for gather_nd

cyrilchim reviewed (commented):
This is good but I would rather remove `tf.gather_nd` with tf.where (let me know if you have any issues with that)

amitvarshney00 committed:
Rename DayCountBasis to DayCountConvention. Remove batching from daycount conventions.

PiperOrigin-RevId: 298376705

pshmakov committed:
Remove usages of tf.TensorArray.

PiperOrigin-RevId: 298584026

tf-quant-finance-robot committed:
Merge pull request #25 from gavincyi:feature/add-daycount-30-360

PiperOrigin-RevId: 298815252

tf-quant-finance-robot committed:
Merge pull request #26 from esztermarton:master

PiperOrigin-RevId: 298823865

pshmakov committed:
Add an XLA test for GenericItoProcess

PiperOrigin-RevId: 298873911

cyrilchimisov-ai committed:
Factor out `results` updates in Monte Carlo scheme.

PiperOrigin-RevId: 298907917

matejr committed:
Bump pip package version to 0.0.1dev18.

PiperOrigin-RevId: 299339701

cyrilchim commented on tf_quant_finance/black_scholes/approximations/american_option.py:
```diff
@@ -0,0 +1,305 @@
+
+
+def _npdf(x):
+  return tf.exp(-0.5 * x ** 2) / tf.math.sqrt(2 * np.pi)
```
tf.math.sqrt(2 * np.pi) is always of dtype tf.float64. Could you please define a global constant like _SQRT_2?

cyrilchim commented on tf_quant_finance/black_scholes/approximations/american_option_test.py:
```diff
@@ -0,0 +1,164 @@
+@test_util.run_all_in_graph_and_eager_modes
+class AmericanPrice(parameterized.TestCase, tf.test.TestCase):
+  """Tests for methods for the american pricing module."""
+
```
Could you please add a dtype as a parameter to at least one of your test?

cyrilchim reviewed (commented):
Looks amazing! Thank you, Rebeka. Just two minor comments. Could you please push it to the repo once done?

cyrilchimisov-ai committed:
Update basic TensorFlow training.

PiperOrigin-RevId: 300057168

tf-quant-finance-robot committed:
Define package's __version__ in terms of segments.

This is in preparation for being able to correctly version automated builds.

PiperOrigin-RevId: 300112075

amitvarshney00 committed:
Interface and pricing for overnight index linked futures (such as OIS, SOFR, SONIA futures etc).

PiperOrigin-RevId: 300177367
